README.md
# Gist

## Installation with Docker (works only locally)

Docker container allows you to run your development environment completely in the background.

Currently, `docker-compose` starts three containers:
 - `web` that holds node.js web application
 - `mongodb` that runs MongoDB
 - `redis` that runs Redis service

 Internally, docker setup separate network with own DNS so it is possible to use container names as the hostnames, e.g. http://web:3000/ URI is accessible from mongodb and redis containers.
 The containers expose defined ports outside on the host machine, e.g, web container expose port 9080 so you can
 open http://localhost:9080/ on your local web browser.

### O. Get the Docker, docker-compose
  - Mac OS: https://docs.docker.com/docker-for-mac/  (`docker-compose` is already included)
  - Ubuntu:
      - https://docs.docker.com/install/linux/docker-ce/ubuntu/
      - https://docs.docker.com/compose/install/#install-compose

### I. Download repo

1. First, clone this repository to your computer with `git clone git@github.com:schemadesign/gist.git`

### II. Creating your `.env.development`, `.env.production` and `.env.testing` files

Environment-related secrets such as the production password database are not committed to this repository as a security-related best practice.

In order to add them (and to support both local development and production deployment) you need to create two files named ".env.development" and ".env.production" in 'config/env/' of this repository at `gist/config/env/.env*` and then fill them respectively with the content which you can find in the EXAMPLE.env. It is necessary to fulfil S3 credentials (first four rows) either from the AWS S3 or from the Digital Ocean Space, also Mapbox to make world map working.

### III. Build the docker container with the web application

1. Run `make develop-build`

### IV. Adding views to the database.

1. Open `seed.json` and make sure it is configured with your local database name
2. Execute `make develop-seed-db` -- the data in `seeds/views.json` will be saved into your database.

Edit `seeds/views.json` when you want to edit the views config, then execute `make develop-seed-db` to save into the database. Commit `seeds/views.json` to Git so everyone can use your updated views.

### V. Running the front-end server locally

#### For Development
1. Run `make develop-start` to start development server in non-daemon mode with logs.

### VI. Changing your localhost settings on your computer and creating a local Gist account

1. Add app.local.arrays.co:
  * Run `sudo nano /etc/hosts` - You should see some numbers including your local ip address on the left column and host names on the right.
  * Add a new line `127.0.0.1    app.local.arrays.co` for the frontend
  * Add another new line `127.0.0.1    local.arrays.co` for the mongodb connection
  * Save the write changes with "control" + "o"
  * Hit "enter"
  * Exit with "control" + "x"

## VII. Creating an user

After the app is up and running you will want to create a user:

1. Access http://app.local.arrays.co:9080/signup/email?superUser=true (If you don't see a form please make command `make develop-grunt`)
2. Fill the form in and add a subdomain

    1. *The subdomain must be listed in the hosts file* (more details in point IX)
    2. email must be *youremail+superuser@schemadesign.com* (if you don't have an email address in the schemadesign.com domain then please update the domain in the code to one of your choices)


3. After confirming you will see an email template in the logs of the backend. Look for an URL with a token.
4. Copy and paste the URL to activate your user.
5. Now you can login in Gist.

### VIII. Add team subdomain to hosts

1. Add your subdomain to hosts
  * Run `sudo nano /etc/hosts` again.
  * Add a new line `127.0.0.1    [your team subdomain].local.arrays.co`
  * Save the write changes with control + "o"
  * Hit "enter"
  * Exit with "control" + "x"
  * You will now be able to navigate to your subdomain at [your team subdomain].local.arrays.co:9080
Anytime you add a new account in Gist, you will have a new subdomain. In order for that subdomain to work locally, you have to add it to your hosts file by repeating step 3.

## Installation on Digital Ocean

### I. Before you upload your code

1. Digital Ocean configuration:
- Create a droplet with Docker
- Create or have a domain (Gist uses subdomains and it won't work on the IP address) and add the Digital Ocean Network with DNSs which direct to your Droplet:
  * `<your_domain>`
  * `app.<your_domain>`
  * `*.<your_domain>`
- Generate new API key in the API tab for the spaces
- Create spaces and add CORS privileges for origin `http(s)://app.<your_domain>` for all methods and add Header `*`.

2. Create a Mapbox account.
Gist is using Mapbox to render world map.
- Create access key
- Create a basic layout to show a map containing countries only
- Create a roads layout to show a map containing roads only

3. Add files and changes in the code
- Find all occurrences of `gist.info` in the code and replace with your domain
- To create the first user we need to configure emails in the file `nodemailer.js` to get activation email or we will see this email on docker logs, but we need to change all occurrences of `@schemadesign.com` email to your email domain.
- Add `env.production` file in the `config/env` using `Exmaple.env` as an example and replace:
    * DO_ACCESS_KEY_ID=(digital ocean key id)
    * DO_SECRET_ACCESS_KEY=(digital ocean access key)
    * DO_S3_BUCKET=(digital ocean space name)
    * DO_S3_ENDPOINT=(digital ocean space endpoint for example `nyc3.digitaloceanspaces.com`, also please find this url in the code and replace to your endpoint)
    * MAPBOX_ACCESS_KEY=(your Mapbox access token)
    * MAPBOX_BASIC_LAYOUT=(URL of the basic layout)
    * MAPBOX_ROADS_LAYOUT=(URL of the basic layout)
    * FONT_CENTRANO_CDN=(URL to the Centra No font, or you can leave it empty and put the font in the folder `public/fonts/centrano2`. You can also replace font in the `public/stylesheets/base/typography/font-faces.css` and later replace all occurrences of the centrano2)
    * HOST=(your domain)
    * SUBDOMAIN=(your main subdomain)
    * NODE_ENV=production
    * MONGODB_URI=mongodb://*login*:*password*@mongodb:27017/arraysdb?authSource=admin&readPreference=primary&appname=MongoDB%20Compass&ssl=false (please replace login and password with you credentials which will be used by the admin user in mongodb)
    * USE_SSL=(`true` if your domain has SSL connection, `false` if your domain doesn't have SSL)
    * PORT=(`443` if your domain has SSL connection, `80` if your domain doesn't have SSL)

### II. Deploy your code

1. Connect to your droplet for example via ssh and deploy code there.
2. Build the app by running commands:
- `make build`
- `make develop-seed-db`
- `make start`
3. Configure MongoDB
- open the terminal and enter `mongodb` container with command `docker-compose exec  -e COLUMNS=222 mongodb bash`
- enter mongo shell with command `mongo`
- switch to admin db with command `use admin`
- create admin user with command `db.createUser({user: "your-login", pwd: "your-password", roles: [ { role: "readWrite", db: “admin”} ] })` (please replace login and password with your credentials which you put in the `MONGODB_URI`)
4. Restart the app with commands `make stop` and `make start`

### III. Creating the first user
1. Enter URL `http://app.<your-domain>/signup/email?superUser=true` and fill the data (if you don't see anything please run the command `make develop-grunt`). Please use as email `+superuser@schemadesign.com` or if you replaced to your email domain, with your domain `+superuser@<your-domain>`
2. In the docker logs you should see an email activation with the URL to activate account. If not you can also activate a user in the mongo shell by changing `activated` to `true`.


#### Or use dnsmasq

Map *.local.arrays.co to localhost, so new teams and subdomains do not have to be manually added to `/etc/hosts` every time.

Install dnsmasq:
`brew install dnsmasq`

Add *.local.arrays.co to config file:
`echo 'address=/.local.arrays.co/127.0.0.1' > $(brew --prefix)/etc/dnsmasq.conf`

Add to resolvers:
`sudo bash -c 'echo "nameserver 127.0.0.1" > /etc/resolver/local.arrays.co'`

if you do not have `/etc/resolver`, simply create it by `mkdir /etc/resolver`

Start dnsmasq and run on boot:
`sudo brew services start dnsmasq`

---------------------

## Running tests

---------------------

1. Run `make test`

This will run all fronted and backend tests in the container.

2. Gist Integration tests are kept [here](https://github.com/schemadesign/gist_integration_tests)

1. Run `npm install` to prepare all files and then
2. Run `npm run test:backend` for backend tests and `npm run test:frontend` for frontend tests.
-----------------------

## All make commands

-----------------------

Make commands in `development` environment

 - `develop` Run development services
 - `develop-start` Run development services & logs
 - `develop-status` Show the status of services.
 - `develop-logs` Show the logs on the terminal from web container.
 - `develop-stop` Stop the containers.
 - `develop-restart` Restart the containers.
 - `develop-build` Rebuild web container.
 - `develop-build-nocache` build web container (no cache)
 - `develop-restart-node` restart node inside the container
 - `develop-seed-db` Seed database (`npm run seed`).
 - `develop-exec-bash` Run bash in the web container.

Make commands for testing

 - `test` run tests in `testing` enviroment

Misc commands

 - `run-debugger` stop the development web container and run again with debug mode (see below)
 - `stop-all` stop all container
 - `lint-fix` for running JS linter.

---------------------
## Run app in debugger mode.

1. `make run-debugger`

This will stop the `web` container and run it again, exposing debugger port `9229` and run in non daemon mode (all logs in the terminal) with a single worker.
Keep in mind if you run your app in the debugger mode, you cannot use other `make develop-*` commands.
You can stop the debug-mode by `Ctrl+C`;

2. Click `Debug` button in VS code.
3. Set the `configuration`:

```json
        {
            "type": "node",
            "request": "attach",
            "name": "Docker: Attach to Node",
            "port": 9229,
            "address": "localhost",
            "localRoot": "${workspaceFolder}",
            "remoteRoot": "/src",
            "protocol": "inspector"
        }
```


### Issue `Unable to open devtools socket: address already in use`

This is usually related to the fact that two node instances tries to bind the same port. The simple solution is to stop the command and run it again with Ctrl+C;
You can also check if the `web` container is running with `docker ps` command.

---------------------

## Troubleshooting

---------------------

1. The Gist app loads AWS credentials from the .env files. There are quite a few ways that AWS credentials can be loaded, and they exist in a *provider chain*. Keep this in mind if you're running into AWS related issues: you may be unintentionally overriding the credentials.
2. If you are seeing errors related to the `sharp` node package, try this:
    * Change directory (`cd [the path to]/gist/node_modules/sharp`) into the `sharp` directory of your local clone of this repository
    * Run `npm install`
3. By default the mongodb and redis containers exposes locally the ports (27017 and 6379 respectively). If you try to start again containers, you get the following error:

    `ERROR: for mongodb  Cannot start service mongodb: driver failed programming external connectivity on endpoint arraystesting_mongodb_1 (0d0b754cd4c943ccdc051b14a38cde0a11c65409f9d3916395a1905d55ebd024): Bind for 0.0.0.0:27017 failed: port is already allocated`

    You have to kill the containers that already use the ports, e.g. `docker kill $(docker ps -q)` to kill all containers.

    By default the testing containers does not expose ports to mongodb and redis. However, you can add this by editing `docker-compose.test.yml` and enabling lines with `- ports:`.

4. Problem with existing container name:

    `ERROR: for mongodb  Cannot create container for service mongodb: Conflict. The container name "/mongodb" is already in use by container "86e6470e1e7dd82e798fa677cfc210c4aac0d749e68f5907123f6c005fea71b2". You have to remove (or rename) that container to be able to reuse that name.`

    Check the containers that runs already by `docker ps` and kill container with the same name.

5. Travis stops because of wrong encryption key

    If your travis-ci build returns following error:

    `$ openssl aes-256-cbc -K $encrypted_29d90a842206_key -iv $encrypted_29d90a842206_iv -in .env.testing.enc -out config/env/.env.testing -d
    bad decrypt
    140444667311776:error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt:evp_enc.c:539:
    The command "openssl aes-256-cbc -K $encrypted_29d90a842206_key -iv $encrypted_29d90a842206_iv -in .env.testing.enc -out config/env/.env.testing -d"
    `

    then you have to recreate `.env.testing.inc` file that holds the content of `config/env/.env.testing`

        1. Run `travis encrypt-file config/env/.env.testing`
        2. Edit `.travis.yml` and replace line in `before_install:` which starts of `- openssl` with the line returned by the `travis` command.

6. Running or building container cause unexpected errors related to wrong .json files

    1. Run the build process without cache `make develop-build-nocache` or `make test-build-nocache`.

7. Travis fails with a nonsense error - how to live?

    This can happen from time to time. If your tests run correctly on your local machine (via `make test` command) then you have to debug the problem directly in the travis service.
    To do that you can use **Debug build** available on the page with the task.
    Reference: https://docs.travis-ci.com/user/running-build-in-debug-mode/

8. Clean up your local docker environment

    You can use script `bin/cleanup-docker.sh`.

9. Error from trying to seed db:

    `err =  { MongoError: failed to connect to server [undefined:27017] on first connect [MongoError: getaddrinfo ENOTFOUND undefined undefined:27017]`

    include the node environment in run seed command `NODE_ENV=development npm run seed`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.9.8] - 2020-09-10
### Fixed

-   Fixed showing revision detail view [#3148]
-   Fixed url of show all button in grouped gallery when group by is set to date or number [#3148]

## [1.9.7] - 2020-07-10

### Added

-   Added endpoint to download ancillary file from public datasets for Gist products [#3141]

## [1.9.6] - 2020-07-02

### Changed

-   Allowed get website endpoint to be fetched without login [#3130]
-   Changed displaying delimited fields in table view [#3133]
-   Enabled scroll header on mobile [#3135]
-   Changed breakpoint for collapsing icons in header [#3135]
-   Redesign view tile in views tab [#3138]

### Fixed

-   Fixed redirect in simple pie set [#3132]
-   Fixed font size for percentage in pie chart on firefox [#3134]
-   Fixed displaying secondary title in gallery on firefox [#3134]
-   Fixed resize and size map on mobile [#3135]

### Added

-   Added option in gallery and timeline to toggle displaying field titles [#3136]
-   Added possibility to toggle full-screen option for embed mode [#3137]
-   Added translation for legend copy for dutch and english [#3124]

## [1.9.5] - 2020-06-22

### Added

-   Added Heatmap for map [#3121]
-   Added Markdown and hiperlink for subtitle in content view [#3123]

### Changed

-   Allowed enterprise team to upload json files [#3122]
-   Remove Grouped Gallery [#3127]

### Fixed

-   Fixed remove previous datasets on remove dataset [#3126]
-   Added handler for getting removed user from versions [#3125]

## [1.9.4] - 2020-06-02

### Added

-   Added download link of dataset in views tab [#3113]

### Changed

-   Improve map quality to show small countries and add addtional countries to map on iso 3 [#3117]

### Fixed

-   Fix vertical scrolling on hided legend in pie set [#3116]
-   Fix rounding circle for versioning of dataset [#3118]
-   Fix counting datasets in team popup [#3118]

## [1.9.3] - 2020-05-14

### Changed

-   Change styles for previous versions [#3111]
-   Updated readme [#3112]

### Fixed

-   Fixed showing warning of team without visualization or last update [#3102]
-   Fixed counting y on simple bar chart [#3099]
-   Fixed outline on country map [#3103]
-   Fixed ticks when is date type on cartesian charts [#3106]
-   Fixed team order [#3109]
-   Replaced waring triangle icon [#3110]

## Add

-   Added scale type for bubble map [#3103][#3104]
-   Added last updated & rearrange header expander bar [#3107]

## [1.9.2] - 2020-05-05

### Added

-   Friendlier token expiration message [#3096]

### Changed

-   Change warning label for csv and tsv upload screen [#3096]
-   Change saving processed raw row data [#3096]

### Fixed

-   Fix safari progress bar issue [#3096]

## [1.9.1] - 2020-04-28

### Added

-   Add information for superuser about inactive teams [#3091]
-   Add warning label for csv and tsv upload screen [#3091]

### Changed

-   Update nodemailer and aws-sdk [#3091]
-   Hide db integration for non super users [#3091]
-   Improve processing raw objects [#3091]

### Fixed

-   Fix searching and filtering in pie chart [#3091]
-   Fix replace copy timeline to grouped-gallery in url [#3091]

## [1.9.0] - 2020-04-20

### Added

-   Allow to register user with token [#3088]

## [1.8.10] - 2020-04-15

### Added

-   Add additional attempts for automatic update fails and add possibility to remove entirely job [#3085]
-   Add Field names to Gallery and Grouped Gallery View [#3085]

### Fixed

-   Fix downloading json file [#3085]

## [1.8.9] - 2020-03-26

### Added

-   Add Segment By to Pie Chart [#3079]
-   Add Segment By to Pie Set [#3079]
-   Add Segment By to Treemap [#3079]
-   Add Segment By to Line Chart [#3079]
-   Add Segment By to Area Chart [#3079]
-

## [1.8.8] - 2020-03-19

### Fixed

-   Fix Portugal and Spain translation for table [#3075]
-   Fix formatting data, number and percent in heatmap [#3075]

### Added

-   Add tile display for heatmap as checkbox [#3075]

### Changed

-   Uid as a param for dataset download endpoint [#3073]

## [1.8.7] - 2020-03-10

### Added

-   Add excluding fields to treemap [#3071]

### Fixed

-   Fix showing percents in pie set [#3071]
-   Fix excluding fields in pie set [#3071]

## [1.8.6] - 2020-03-02

### Fixed

-   Prevent creating empty delimited field on editing content [#3069]

## [1.8.5] - 2020-02-26

### Fixed

-   Fix slider filter [#3067]

## [1.8.4] - 2020-02-20

### Fixed

-   Fix filter with delimited fields which could be empty [#3064]

## [1.8.3] - 2020-02-13

### Added

-   Add percentage to pie chart/set [#3060]

## [1.8.2] - 2020-02-12

### Changed

-   Website API get modified to be accessible without credentials [#3056]

### Fixed

-   Restore Restore sentence splitting to words in wordcloud [#3050]
-   Added formatting fix for simple pie set [#3052]
-   Filter off invalid dates in bubble chart timeline [#3053]
-   Fix save button [#3055]
-   Fix going back [#3055]

## [1.8.1] - 2020-02-11

### Fixed

-   Fix creating delimited fields [#3057]

## [1.8.0] - 2020-01-30

### Added

-   Migrated to Mongo 3.6 and Mongoose 5.8 [#3038]
-   Added insight explorer to embeded mode [#3045]

## [1.7.17] - 2020-01-24

### Added

-   Added insight explorer to embeded mode

### Fixed

-   Fixed precise on axis on cartesian charts [#3038]
-   Fixed page copying [#3046]

## [1.7.16] - 2020-01-16

### Fixed

-   Fix calculation by same category on simple Cartesian chart [#3040]

## [1.7.15] - 2019-12-09

### Fixed

-   Fix displaying insight explorer on creating first insight [#3034]

## [1.7.14] - 2019-12-06

### Fixed

-   Restyle insights explorer [#3030]
-   Fix simple cartesian charts with missing meta data [#3030]
-   Always set insight as public on creating [#3031]

### Added


app.js
const winston = require('winston');
const dotenv = require('dotenv');
const cluster = require('cluster');
const path = require('path');
const fs = require('fs');
const async = require('async');
const http = require('http');
const https = require('https');

const { configureWinston, expressLogger } = require('./app/utils/winston');

dotenv.config({
    path: `${__dirname}/config/env/.env.${process.env.NODE_ENV || 'development'}`,
    silent: true,
});

const isDev = process.env.NODE_ENV === 'development';

var app;

var userFolderPath = __dirname + '/user';
global.rootPath = __dirname;

configureWinston();

function module_exists(name) {
    try {
        return require.resolve(name);
    } catch (e) {
        winston.debug(`module_exists error: ${e.message}`);
        return false;
    }
}

const clusterWorkerSize = parseInt(process.env.WEB_CONCURRENCY || require('os').cpus().length, 10);

winston.info('Cluster worker size: ' + clusterWorkerSize);

var mongoose_client = require('./app/models/mongoose_client');

if (cluster.isMaster && !['testing', 'test'].includes(process.env.NODE_ENV) && clusterWorkerSize > 1) {

    var cron = require('./app/boot/cron-init.js');

    var worker;
    winston.debug('env %s, master pid %s, total Workers: %s ', process.env.NODE_ENV, process.pid, clusterWorkerSize);

    for (var i = 0; i < clusterWorkerSize; i++) {
        worker = cluster.fork();

        worker.on('message', function (msg) {

            cron.execute(msg.command, msg.id, function (err) {

                if (err) {
                    winston.error('err in delegating work to master node : ' + err);
                } else {
                    winston.debug('delegate schedule work to master node **** command: ' + msg.command + ', cronJobId: ' + msg.id);
                }

            });
        });
    }

    cluster.on('exit', function (worker_process) {
        worker = cluster.fork();
        winston.debug('Worker %d died pid %d', worker_process.id, process.pid);

    });

    /**
     * Run custom boot scripts, if present
     */
    fs.readdir(userFolderPath, function (err, files) {
        async.each(files, function (file, eachCb) {
            var full_path = path.join(userFolderPath, file);

            fs.stat(full_path, function (err, stat) {
                if (err) {
                    eachCb(err);
                } else {
                    if (stat.isDirectory()) {
                        var customBoot = path.join(full_path, '/app/boot.js');

                        if (module_exists(customBoot)) {
                            try {
                                require(customBoot);
                                winston.info('Loaded custom boot file for: %s', file);
                            } catch (e) {
                                winston.error('Error loading custom boot file for: %s\n%s', file, e);
                            }
                        }
                    }
                    eachCb();
                }
            });
        });
    });

} else {

    var express = require('express');

    var cookieParser = require('cookie-parser');
    var bodyParser = require('body-parser');
    var expressValidator = require('express-validator');
    var session = require('express-session');
    var MongoSessionStore = require('connect-mongo')(session);
    var flash = require('connect-flash');
    var passport = require('passport');
    var cors = require('cors');
    var subdomain = require('express-subdomain');
    var composableMiddleware = require('composable-middleware');
    const serveFavicon = require('serve-favicon');
    const helmet = require('helmet');

    const authRoute = require('./app/routes/auth');
    const accountRoute = require('./app/routes/account');

    app = express();
    app.enable('trust proxy');

    require('./config/setup-passport');

    var customRoutes = __dirname + '/app/routes/custom';

    var viewsToSet = [];
    var userNunjucksPath = [];

    viewsToSet.push(__dirname + '/views');

    const expressNunjucks = require('express-nunjucks');
    const { FileSystemLoader } = require('nunjucks');
    app.set('view engine', 'njk');

    const loadNunjucks = () => {
        const njk = expressNunjucks(app, {
            watch: isDev,
            noCache: isDev,
            loader: FileSystemLoader,
        });
        require('./nunjucks/filters')(njk.env, process.env);
        return njk;
    };

    if (process.env.NODE_ENV !== 'enterprise') {
        app.set('subdomain offset', process.env.HOST.split('.').length);
    }

    app.use(helmet({
        frameguard: false,
    }));

    fs.readdir(userFolderPath, function (err, files) {

        if (!files) {
            app.set('views', viewsToSet);
            return loadNunjucks();
        }

        files = files.filter(function (item) {
            return !(/(^|\/)\.[^/.]/g).test(item);
        });

        async.each(files, function (file, eachCb) {
            var full_path = path.join(userFolderPath, file);
            var team_name = file;
            fs.stat(full_path, function (err, stat) {
                if (err) {
                    eachCb(err);
                } else {
                    if (stat.isDirectory() && files) {
                        var nunjucksFilePath = path.join(userFolderPath, file, 'nunjucks/filters.js');
                        try {
                            fs.statSync(nunjucksFilePath);
                            userNunjucksPath.push(nunjucksFilePath);
                        } catch (e) {
                            // winston.error('no filters path for ' + userFolderPath);
                        }
                        //serving the views
                        var view_path = path.join(userFolderPath, file, '/views');
                        var static_path = path.join(userFolderPath, file, '/static');
                        app.use('/' + team_name, express.static(static_path));
                        viewsToSet.push(view_path);

                        var customApi = path.join(userFolderPath, file, '/src/customApi.js');

                        // Only throw errors for customApi modules that exist
                        if (module_exists(customApi)) {
                            try {
                                var middleware = composableMiddleware().use(require(customApi)).use(require(customRoutes));
                                app.use('/api', require('./app/routes/api'));
                                app.use('/auth', helmet.frameguard(), authRoute);
                                app.use('/account', helmet.frameguard(), accountRoute);
                                app.use(subdomain(team_name, middleware));
                            } catch (e) {
                                winston.debug(`app.use error: ${e.message}`);
                                app.use(subdomain(team_name, require(customRoutes)));
                            }
                        } else {
                            app.use(subdomain(team_name, require(customRoutes)));
                        }
                    }
                    eachCb();
                }
            });

            // Set environment variables for custom teams that were loaded
            process.env[team_name.toUpperCase()] = true;
        }, function (err) {
            if (err) {
                return winston.error('cannot sync the user folder files :', err);
            }

            app.set('views', viewsToSet);

            const njk = loadNunjucks();
            userNunjucksPath.forEach(path => require(path)(njk.env, process.env));
        });
    });

    app.use('/**/fonts/*', cors());
    app.use('/**/map-pin.png', cors());

    const favicon = process.env.INSIGHT ? path.join(__dirname, '/public/images/insight-favicon.png') :
        path.join(__dirname, '/public/images/favicon.ico');

    app.use(serveFavicon(favicon));
    app.use(express.static(path.join(__dirname, '/public')));

    var routes = require('./app/routes');

    app.use(bodyParser.urlencoded({ extended: false, limit: '50mb' })); // application/x-www-form-urlencoded
    app.use(bodyParser.json({ limit: '50mb' })); // application/JSON
    app.use(expressValidator()); // this line must be immediately after any of the bodyParser middlewares!
    app.use(require('compression')());
    app.set('trust proxy', true);
    app.use(cookieParser());

    var domain = 'localhost';

    if (process.env.HOST) {
        var urlParts = process.env.HOST.split('.');
        urlParts.splice(0, urlParts.length - 2);
        // Remove port
        urlParts[urlParts.length - 1] = urlParts[urlParts.length - 1].split(':')[0];
        domain = '.' + urlParts.join('.');
    }

    const cookiesPolicy = process.env.USE_SSL === 'true' ? {
        secure: true,
        httpOnly: true,
    } : {};

    // Mongo Store to prevent a warning.
    app.use(session({
        secret: process.env.SESSION_SECRET,
        rolling: true,
        resave: false,
        saveUninitialized: false,
        name: 'gist-session-nam3',
        cookie: {
            domain: domain,
            // 60 minutes unless otherwise specified
            maxAge: +process.env.SESSION_TIMEOUT || (1000 * 60 * 60),
            ...cookiesPolicy,
        },
        store: new MongoSessionStore({
            url: mongoose_client.dbURI,
            // url: process.env.MONGODB_URI ? process.env.MONGODB_URI : 'mongodb://localhost/arraysdb',
            // touchAfter: 240 * 3600 // time period in seconds
        }),
    }));

    app.use(flash());
    app.use(passport.initialize());
    app.use(passport.session());

    // Logger only to use for development
    expressLogger(app);

    var raw_source_documents = require('./app/models/raw_source_documents');

    process.on('unhandledRejection', ({ message }) => {
        winston.error(`Node detected an unhandled rejection! err: ${message}`);
    });

    var modelNames = [raw_source_documents.ModelName];
    mongoose_client.FromApp_Init_IndexesMustBeBuiltForSchemaWithModelsNamed(modelNames);

    mongoose_client.WhenMongoDBConnected(function () {
        mongoose_client.WhenIndexesHaveBeenBuilt(function () {

            // winston.info(" Proceeding to boot app. ");
            //
            routes.MountRoutes(app);

            //
            // Run actual server
            if (module === require.main) {
                let server;

                var privateKeyPath = 'config/ssl/letsencrypt/privkey.pem';
                var certificatePath = 'config/ssl/letsencrypt/fullchain.pem';

                // Create a HTTPS server if the credentials exist
                if (fs.existsSync(privateKeyPath) && fs.existsSync(certificatePath) && process.env.USE_SSL === 'true') {
                    const sslOptions = {
                        key: fs.readFileSync(privateKeyPath, 'utf8'),
                        cert: fs.readFileSync(certificatePath, 'utf8'),
                    };
                    server = https.createServer(sslOptions, app);

                    // Create lightweight server for redirecting http requests to https
                    http.createServer(function (req, res) {
                        res.writeHead(301, { Location: 'https://' + req.headers['host'] + req.url });
                        res.end();
                    }).listen(80);

                    process.env.USE_SSL = 'true';
                } else {
                    server = http.createServer(app);
                }

                server.listen(process.env.PORT || 9080, function () {
                    const host = isDev ? 'localhost' : server.address().address;
                    const port = server.address().port;
                    if (process.env.NODE_ENV === 'testing' || clusterWorkerSize === 1) {
                        winston.debug('listing at %s:%s', host, port);
                    } else {
                        winston.debug('Worker %d running pid %d listing at %s:%s', cluster.worker.id, process.pid, host, port);
                    }
                });
                server.timeout = 30000;
            }
        });
    });
}
module.exports = app;
